File: src/types/rig.ts
// ============================================================================
// TYPES: Areas y Rigs (Taladros)
// ============================================================================

// ─── Area Types ─────────────────────────────────────────────────────────────

export interface Area {
  id: string;
  name: string;
  country: string;
  state: string;
  active: boolean;
  createdBy?: string;
  updatedBy?: string;
  createdAt: string;
  updatedAt: string;
}

export interface CreateAreaInput {
  name: string;
  country: string;
  state: string;
  active: boolean;
}

export interface UpdateAreaInput {
  name?: string;
  country?: string;
  state?: string;
  active?: boolean;
}

// ─── Rig Contractor Types ────────────────────────────────────────────────────

export interface RigContractor {
  id: string;
  rigId: string;
  companyId: string;
  createdAt: string;
}

export interface RigContractorWithCompany extends RigContractor {
  companyName: string;
}

/** Embedded in RigFull — contractor entry joined with company info */
export interface RigContractorEntry {
  rigContractorId: string;
  companyId: string;
  companyName: string;
}

export interface AddRigContractorInput {
  companyId: string;
}

// ─── Rig (Taladro) Types ─────────────────────────────────────────────────────

export interface Rig {
  id: string;
  name: string;
  /** Legacy text field — kept for backwards-compat */
  operator: string;
  operatorId?: string;
  power: string;
  areaId?: string;
  active: boolean;
  createdBy?: string;
  updatedBy?: string;
  createdAt: string;
  updatedAt: string;
}

/** Rig joined with area and operator name — used in list views */
export interface RigWithArea extends Rig {
  operatorName?: string;
  areaName?: string;
  areaCountry?: string;
  areaState?: string;
}

/** Full rig data including contractor list — used by the wizard */
export interface RigFull extends RigWithArea {
  contractors: RigContractorEntry[];
}

export interface CreateRigInput {
  name: string;
  /** Legacy text — still sent for backwards-compat */
  operator: string;
  /** Required FK — must be a valid company id with type 'operator' */
  operatorId: string;
  power: string;
  areaId: string;
  active?: boolean;
  /** At least one contractor id required */
  contractorIds: string[];
}

export interface UpdateRigInput {
  name?: string;
  operator?: string;
  operatorId?: string;
  power?: string;
  areaId?: string;
  active?: boolean;
}

// ─── Rig Personnel (Cuadrilla asociada al taladro) ───────────────────────────

export interface RigPersonnel {
  id: string;
  rigId: string;
  name: string;
  ci?: string;
  defaultPosition: string;
  active: boolean;
  createdAt: string;
  updatedAt: string;
}

export interface CreateRigPersonnelInput {
  name: string;
  ci?: string;
  defaultPosition: string;
}

export interface UpdateRigPersonnelInput {
  name?: string;
  ci?: string;
  defaultPosition?: string;
  active?: boolean;
}

// ─── UI Helper Types ─────────────────────────────────────────────────────────

export interface AreaOption {
  value: string;
  label: string;
  country?: string;
  state?: string;
}

export interface RigOption {
  value: string;
  label: string;
  operatorName?: string;
  areaName?: string;
}

export interface AreaFilters {
  includeInactive?: boolean;
  country?: string;
  state?: string;
  searchTerm?: string;
}

export interface RigFilters {
  includeInactive?: boolean;
  areaId?: string;
  operatorId?: string;
  searchTerm?: string;
}

// ─── Constants ───────────────────────────────────────────────────────────────

export const AREA_REQUIRED_FIELDS = ['name', 'country', 'state'] as const;
export const RIG_REQUIRED_FIELDS = ['name', 'operatorId', 'areaId', 'power', 'contractorIds'] as const;

export const COMMON_COUNTRIES = [
  'Venezuela',
  'Colombia',
  'Ecuador',
  'Perú',
  'Bolivia',
  'Argentina',
  'Chile',
  'Brasil',
  'México',
  'Estados Unidos',
  'Canadá',
] as const;

export const VENEZUELA_STATES = [
  'Anzoátegui',
  'Apure',
  'Barinas',
  'Bolívar',
  'Carabobo',
  'Cojedes',
  'Delta Amacuro',
  'Falcón',
  'Guárico',
  'Lara',
  'Mérida',
  'Miranda',
  'Monagas',
  'Nueva Esparta',
  'Portuguesa',
  'Sucre',
  'Táchira',
  'Trujillo',
  'Vargas',
  'Yaracuy',
  'Zulia',
  'Distrito Capital',
] as const;

File: src-tauri/src/models/rig.rs
use serde::{Deserialize, Serialize};

// ============================================================================
// Core structs
// ============================================================================

#[allow(dead_code)]
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Rig {
    pub id: String,
    pub name: String,
    /// Legacy text field — kept for backwards-compat, prefer operator_id
    pub operator: String,
    pub operator_id: Option<String>,
    pub power: String,
    pub area_id: Option<String>,
    pub active: bool,
    pub created_by: Option<String>,
    pub updated_by: Option<String>,
    pub created_at: String,
    pub updated_at: String,
}

/// Rig joined with its area, operator company, and contractor list
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RigFull {
    pub id: String,
    pub name: String,
    pub operator: String,
    pub operator_id: Option<String>,
    pub operator_name: Option<String>,
    pub power: String,
    pub area_id: Option<String>,
    pub area_name: Option<String>,
    pub area_country: Option<String>,
    pub area_state: Option<String>,
    pub active: bool,
    pub created_at: String,
    pub updated_at: String,
    /// Populated separately — list of contractors assigned to this rig
    pub contractors: Vec<RigContractorEntry>,
}

/// A single contractor entry embedded in RigFull
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RigContractorEntry {
    pub rig_contractor_id: String,
    pub company_id: String,
    pub company_name: String,
}

/// Kept for backwards compat where only area join is needed
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RigWithArea {
    pub id: String,
    pub name: String,
    pub operator: String,
    pub operator_id: Option<String>,
    pub operator_name: Option<String>,
    pub power: String,
    pub area_id: Option<String>,
    pub area_name: Option<String>,
    pub area_country: Option<String>,
    pub area_state: Option<String>,
    pub active: bool,
    pub created_at: String,
    pub updated_at: String,
}

// ============================================================================
// Input structs
// ============================================================================

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CreateRigInput {
    pub name: String,
    /// Legacy — kept so existing callers don't break
    pub operator: String,
    /// New FK field — required for the wizard flow
    pub operator_id: Option<String>,
    pub power: String,
    pub area_id: Option<String>,
    pub active: Option<bool>,
    /// Contractor company IDs to link via rig_contractors on creation
    pub contractor_ids: Option<Vec<String>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UpdateRigInput {
    pub name: Option<String>,
    pub operator: Option<String>,
    pub operator_id: Option<String>,
    pub power: Option<String>,
    pub area_id: Option<String>,
    pub active: Option<bool>,
}
